packages/shared/lib/services/connection.service.ts

```diff
@@ -467,6 +470,50 @@ class ConnectionService {
         return [{ connection: connection!, operation: connection ? 'override' : 'creation' }];
     }
 
+    public async upsertSignatureBasedConnection({
+        connectionId,
+        providerConfigKey,
+        credentials,
+        connectionConfig,
+        metadata,
+        config,
+        environment,
+        account
+    }: {
+        connectionId: string;
+        providerConfigKey: string;
+        credentials: SignatureBasedCredentials;
+        connectionConfig?: ConnectionConfig;
+        config: ProviderConfig;
+        metadata?: Metadata | null;
+        environment: DBEnvironment;
+        account: DBTeam;
+    }): Promise<ConnectionUpsertResponse[]> {
+        const encryptedConnection = encryptionManager.encryptConnection({
+            connection_id: connectionId,
+            provider_config_key: providerConfigKey,
+            config_id: config.id as number,
+            credentials,
+            connection_config: connectionConfig || {},
+            environment_id: environment.id,
+            metadata: metadata || null
+        });
+
+        const [connection] = await db.knex
+            .from<StoredConnection>(`_nango_connections`)
+            .insert(encryptedConnection)
+            .onConflict(['connection_id', 'provider_config_key', 'environment_id', 'deleted_at'])
+            .merge({
+                ...encryptedConnection,
+                updated_at: new Date()
+            })
+            .returning('*');
+
+        void analytics.track(AnalyticsTypes.SIGNATURE_BASED_CONNECTION_INSERTED, account.id, { provider: config.provider });
+
+        return [{ connection: connection!, operation: connection ? 'override' : 'creation' }];
+    }
+
```

Comment:
do we need a new function just for this. All the upsert functions are basically identical

Reply:
I have refactored the upsert functions so that we can have one that will upsert, `BASIC`, `API_KEY`, `JWT`, `BILL`, `TABLEAU`, `SIGNATURE`, `TWO_STEP` and `TBA` connections in the latest [commit](https://github.com/NangoHQ/nango/pull/2942/commits/b4d3dd0ffcd24f1188b251597107f09a4e90eb13)